proposals/nnnn-closure-isolation.md

```diff
@@ -0,0 +1,114 @@
+# Closure isolation control
+
+* Proposal: [SE-NNNN](nnnn-closure-isolation.md)
+* Authors: [Sophia Poirier](https://github.com/sophiapoirier), [John McCall](https://github.com/rjmccall)
+* Review Manager: TBD
+* Implementation: On `main` gated behind `-enable-experimental-feature TODO`
+* Previous Proposals: [SE-0313](0313-actor-isolation-control.md), [SE-0316](0316-global-actors.md)
+* Review: ([pitch](https://forums.swift.org/TODO))
+
+## Introduction
+
+This proposal provides the ability to explicitly specify actor-isolation or non-isolation of a closure, as well as providing a parameter attribute to guarantee that a closure parameter inherits the isolation of the context.
+
+## Motivation
+
+The formal isolation of a closure can be explicitly specified as global actor isolation:
+
+```swift
+Task { @MainActor in
+  print("global actor isolation")
+}
+```
+
+Without a global actor isolation annotation, actor-isolation or non-isolation of a closure is inferred but cannot be explicitly specified. This proposal enables closures to be fully explicit about all three types of formal isolation:
+* `nonisolated`
+* global actor
+* specific actor value
+
+Explicit annotation has the benefit of disabling inference rules and the potential that they lead to a formal isolation that is not preferred. For example, there are circumstances where it is beneficial to guarantee that a closure is `nonisolated` therefore knowing that its execution will hop off the current actor. Explicit annotation also offers the ability to identify a mismatch of intention, such as a case where the developer expected `nonisolated` but inference landed on actor-isolated, and the closure is used in an isolated context. With explicit annotation, the developer would receive a diagnostic about a `nonisolated` closure being used in an actor-isolated context which helpfully identifies this mismatch of intention.
+
+## Proposed solution
+
+Enable explicit specification of non-isolation by allowing `nonisolated` to be a specifier on a closure:
+
+```swift
+Task { nonisolated in
+  print("nonisolated")
+}
+```
+
+Enable explicit specification of actor-isolation via an isolated parameter in a closure's capture list by using the `isolated` specifier:
+
+```swift
+actor A {
+  func isolate() {
+    Task { [isolated self] in
```

Comment:
Does this imply the following scenarios?

```swift
func isolate(on otherActor: some AnyActor) {
  Task { [isolated otherActor] in
      // Valid
  }
}
```

```swift
func isolate(on otherActor: some AnyActor) {
  Task { [isolated otherActor] in
     self.bool.toggle() // Fails compilation because `otherActor` isolation does not match `self`'s isolation
  }
}
```

```swift
protocol Toggleable {
  func toggle()
}

func isolate(on otherActor: some AnyActor & Toggleable) {
  Task { [isolated otherActor] in
    otherActor.toggle() // Successful, because this Task is isolated to `otherActor`'s isolation
  }
}
```

Reply:
(Let me preface by saying that this proposal isn't fully baked yet, hence not having been pitched or added anyone to the PR yet.)

The first example, it is unclear to me what is "valid" given that the body of the closure does not do anything? But the syntax for the isolation capture is valid, and the capture will be isolated to `otherActor`.

The second example correct will fail.

The third example yes correct as well.